riscv/src/small_field/code_gen.rs

```diff
@@ -0,0 +1,1793 @@
+    instr is_greater_or_equal_signed XL, YL, WL
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, XL, STEP)
+        link ~> (tmp2_h, tmp2_l) = regs.mload(0, YL, STEP + 1)
+        link ~> tmp6_l = add_sub.gt(tmp1_h, tmp1_l, tmp2_h, tmp2_l)
+        link ~> regs.mstore(0, WL, STEP + 2, 0, X_b3)
+        link => byte.check(X_b1)
+        link => byte.check(X_b2)
+        link => byte.check(X_b3)
+        link => bit7.check(tmp3_l)
+        link => bit7.check(tmp3_h)
+    {
+        std::utils::force_bool(tmp6_l),
+
+        tmp1_h = X_b1 + tmp3_l * 2**8 + wrap_bit * 2**15,
+        tmp2_h = X_b2 + tmp3_h * 2**8 + wrap_bit_2 * 2**15,
+
+        tmp4_l = (1 - wrap_bit) * (1 - wrap_bit_2),
+        tmp4_h = wrap_bit * wrap_bit_2,
+        tmp5_l = (1 - wrap_bit) * wrap_bit_2,
+        //tmp5_h = wrap_bit * (1 - wrap_bit_2),
+ 
+        X_b3 = tmp4_l * (1 - tmp6_l) +
+               tmp4_h * (1 - tmp6_l) +
+               tmp5_l /* * 1 */
+               /* tmp5_h * 0, */
+    }
+
+    // Stores val(X) * Z + W in register Y.
+    instr affine XL, YL, ZH, ZL, WH, WL
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, XL, STEP)
+        link ~> (tmp3_h, tmp3_l, tmp2_h, tmp2_l) = arith_mul.mul(tmp1_h, tmp1_l, 0, 0, ZH, ZL)
+        // we ignore tmp3 because that's the high 32 bits of the 64 bits multiplication result
+        link ~> (tmp4_h, tmp4_l) = add_sub.add(tmp2_h, tmp2_l, WH, WL)
+        link ~> regs.mstore(0, YL, STEP + 1, tmp4_h, tmp4_l);
+
+    // ================= wrapping instructions =================
+
+    // Computes V = val(X) + val(Y) + Z, wraps it in 32 bits, and stores the result in register W.
+    // Requires 0 <= V < 2**33.
+    instr add_wrap XL, YL, ZH, ZL, WL
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, XL, STEP)
+        link ~> (tmp2_h, tmp2_l) = regs.mload(0, YL, STEP + 1)
+        link ~> (tmp3_h, tmp3_l) = add_sub.add(tmp2_h, tmp2_l, ZH, ZL)
+        link ~> (tmp4_h, tmp4_l) = add_sub.add(tmp1_h, tmp1_l, tmp3_h, tmp3_l)
+        link ~> regs.mstore(0, WL, STEP + 2, tmp4_h, tmp4_l);
+
+    // Computes V = val(X) - val(Y) + Z, wraps it in 32 bits, and stores the result in register W.
+    // Requires -2**32 <= V < 2**32.
+    instr sub_wrap_with_offset XL, YL, ZH, ZL, WL
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, XL, STEP)
+        link ~> (tmp2_h, tmp2_l) = regs.mload(0, YL, STEP + 1)
+        link ~> (tmp3_h, tmp3_l) = add_sub.sub(tmp1_h, tmp1_l, tmp2_h, tmp2_l)
+        link ~> (tmp4_h, tmp4_l) = add_sub.add(tmp3_h, tmp3_l, ZH, ZL)
+        link ~> regs.mstore(0, WL, STEP + 2, tmp4_h, tmp4_l);
+
+    // ================= logical instructions =================
+
+    // Stores 1 in register W if the value in register X is zero,
+    // otherwise stores 0.
+    instr is_equal_zero XL, WL
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, XL, STEP)
+        link ~> regs.mstore(0, WL, STEP + 2, 0, XXIsZero)
+    {
+        XXIsZero = 1 - XX * XX_inv,
+        XX = (tmp1_h + tmp1_l)
```

Comment:
This is sound because `tmp1_h` and `tmp1_l` cannot be opposites of each other?

Reply:
not sure I follow. They are unrelated to each other, but they do compose the 32-bit number `tmp1_h * 2**16 + tmp1_l`. We want to check whether that 32-bit number is 0, which is only true if both limbs are 0

Reply:
My previous comment was wrong, my bad!
Right, and checking the sum works because the limbs cannot be negative?

Reply:
yes each limb is always unsigned 2 bytes. The whole 32-bit number may be negative in 2s complement